### ISSUE TITLE
Pod Logs outside of --data-dir

### ISSUE THREAD

**Issue by pschichtel** [k0sproject/k0s#3508]
### Before creating an issue, make sure you've checked the following:

- [X] You are running the latest released version of k0s
- [X] Make sure you've searched for existing issues, both open and closed
- [X] Make sure you've searched for PRs too, a fix might've been merged already
- [X] You're looking at docs for the released version, "main" branch docs are usually ahead of released versions.

### Platform

```shell
Rocky 9
```

### Version

v1.27.4+k0s.0

### Sysinfo

I'll follow this up if necessary. I can also upgrade the deployment to 1.27.5 if necessary later on.

### What happened?

The k0s deployment has the data-directory configured to be /var/kubernetes, which is on a large partition. The root partition is fairly small. One of our nodes ran out of disk space with most of the usage being in /var/log/pods. I would have assumed that pod logs also go to a subfolder of the data-directory.

### Steps to reproduce

1. Deploy k0s with --data-dir pointing to /var/kubernetes.
2. Deploy a pod, that produces tons of logs
3. Watch the /var/log/pods folder grow in size

### Expected behavior

Logs are inside of data dir

### Actual behavior

Logs are in /var/log/pods

**Comment by jnummelin** [comment]
Did you try with adding `--kubelet-extra-args`, kubelet has some flags to control the log file handling, such as:
* `--container-log-max-files int32     Default: 5`
* `--container-log-max-size string     Default: 10Mi`

Other than that, kubelet does not have flags to scope where the log files are created unfortunately.

One option is to mount a partition (or use bind mounts) for `/var/log/pods` to have them out of the root partition.
**NOTE** I've never actually tried this, so mileage may vary 😄

**Comment by pschichtel** [comment]
I'm pretty sure a bind mount would do the trick, yes. Apparently even symlinks work fine from what I found just now with google.

> Other than that, kubelet does not have flags to scope where the log files are created unfortunately.

That's surprising to me, but then I guess there is not much you can do here. Where would be the right place to create a feature request upstream? The [kubelet](https://github.com/kubernetes/kubelet) repo doesn't seem super active.

**Comment by pschichtel** [comment]
I looked around a bit... The path is defined [here](https://github.com/kubernetes/kubernetes/blob/46835f8792dfb4a17345e592d1325bf63bc054e4/pkg/kubelet/kuberuntime/kuberuntime_manager.go#L73) and it seems to be used in quite a few places. I'm looking into the KEP process, but I guess this will probably not change soon.

Feel free to keep this issue around or close it, as far as k8s is concerned it currently works as intended. Also bind mounts and symlinks are both reasonable "workarounds".

**Comment by pschichtel** [comment]
I wonder though: Could setting up a bind mount be something that k0s does automatically during startup?

**Comment by twz123** [comment]
> I wonder though: Could setting up a bind mount be something that k0s does automatically during startup?

Maybe not k0s itself, but If you're using k0sctl, you might be able to automate this on a per-node basis via [before hooks](https://github.com/k0sproject/k0sctl/tree/v0.16.0#spechostshooks-mapping-optional).

**Comment by pschichtel** [comment]
Yeah makes sense. I guess this covers the issue well enough until upstream supports relocating logs. Feel free to close this.

**Comment by ncopa** [comment]
Seems like it is configurable via `podLogsDir`.

> Kubelet allows changing the pod logs directory from default /var/log/pods to a custom path. This adjustment can be made by configuring the podLogsDir parameter in the kubelet's configuration file.

https://kubernetes.io/docs/concepts/cluster-administration/logging/#log-location-node

**Comment by pschichtel** [comment]
@ncopa is that exposed through k0s' configuration?

**Comment by twz123** [comment]
Wow, the podLogsDir knob has been released as part of Kubernetes 1.30. Definitely missed that in the release notes. It should be configurable right now via worker profiles, I guess. That's k0s's way of managing the kubelets config file.

However, I'd say the issue is still unfixed on k0s side, as k0s should actually default podLogsDir to something inside the data dir.

**Comment by pschichtel** [comment]
I wonder if this would break assumptions of log forwarders like promtail, filebeat, ...

**Comment by jnummelin** [comment]
> I wonder if this would break assumptions of log forwarders like promtail, filebeat, ...

I'm pretty sure it would. Like we've seen with k0s using `/var/lib/k0s/kubelet` is breaking assumptions on some CSI etc providers. See #1842 

But then again, if this is now actually configurable on kubelet, those components should have the path also configurable.

🤷

**Comment by pschichtel** [comment]
Yeah other k8s distros also hit this. The CSI drivers I worked with all already properly supported this, but I'm not sure the common log forwardes are there yet, especially since this is a new feature. Nothing that can't easily fixed in those projects, but there will be issues for sure.

I think out of those that I worked with, I think filebeat is the only forwarder that might support this out of the box. But that's mostly because it is already a major PITA to setup (read: requires a lot of explicit configuration with not a lot documentation to go along with that) anyway, one slight change in configuration won't make it any more annoying.

**Comment by jnummelin** [comment]
> one slight change in configuration won't make it any more annoying

😹

**Comment by jnummelin** [comment]
So now that this is actually configurable on kubelet, we (the maintainers) decided that we'll keep the current defaults still. This is mainly driven by the fact that there's unknown amount of other _things_ that might break as this is pretty new feature in k8s itself. And also changing the defaults in general is bit bad idea.

So instead, we're working towards configuration v2 for whole k0s. With that we will revisit this topic too as changing config versions means we actually can change the defaults. For overall config v2 goals see https://github.com/k0sproject/k0s/pull/4920
